subprojects/build-plugins/src/test/kotlin/org/gradle/kotlin/dsl/build/plugins/GradleApiExtensionsIntegrationTest.kt

```diff
@@ -0,0 +1,189 @@
+
+            tasks.withType<GenerateKotlinDslApiExtensions> {
+                isUseEmbeddedKotlinDslProvider.set(true)
+            }
+
+            tasks.withType<JavaCompile> {
+                options.isFork = true
+            }
+
+            repositories {
+                jcenter()
+            }
+        """)
+
+        run("generateKotlinDslApiExtensions")
+
+        existing("build/generated-sources").walkTopDown().single { it.isFile }.let {
+            Assert.assertThat(it.name, CoreMatchers.equalTo("GeneratedGradleApiExtensionsMainKotlinDslApiExtensions.kt"))
+            val extensions = listOf(
+                """
+                fun <S : T, T : Any> org.gradle.api.DomainObjectSet<T>.`withType`(`type`: kotlin.reflect.KClass<S>): org.gradle.api.DomainObjectSet<S> =
+                    `withType`(`type`.java)
+                """,
+                """
+                fun org.gradle.api.tasks.AbstractCopyTask.`filter`(`filterType`: kotlin.reflect.KClass<java.io.FilterReader>, vararg `properties`: Pair<String, *>): org.gradle.api.tasks.AbstractCopyTask =
+                    `filter`(mapOf(*`properties`), `filterType`.java)
+                """,
+                """
+                @org.gradle.api.Incubating
+                fun <T : org.gradle.api.Task> org.gradle.api.tasks.TaskContainer.`register`(`name`: String, `type`: kotlin.reflect.KClass<T>, `configurationAction`: T.() -> Unit): org.gradle.api.tasks.TaskProvider<T> =
+                    `register`(`name`, `type`.java, `configurationAction`)
+                """,
+                """
+                @Deprecated("Deprecated Gradle API")
+                @org.gradle.api.Incubating
+                fun <T : org.gradle.api.Task> org.gradle.api.tasks.TaskContainer.`createLater`(`name`: String, `type`: kotlin.reflect.KClass<T>, `configurationAction`: T.() -> Unit): org.gradle.api.tasks.TaskProvider<T> =
+                    `createLater`(`name`, `type`.java, `configurationAction`)
+                """)
+            Assert.assertThat(it.readText(), CoreMatchers.allOf(extensions.map { containsMultiLineString(it) }))
+        }
+
+        run("assemble")
+
+        existing("extract").let { extract ->
+            unzipTo(extract, existing("build/libs/gradle-api-extensions.jar"))
+            Assert.assertThat(
+                extract.walkTopDown().filter { it.isFile }.map { it.relativeTo(extract).path }.toList(),
+                CoreMatchers.hasItems(*listOf(
+                    "org/gradle/kotlin/dsl/GeneratedGradleApiExtensionsMainKotlinDslApiExtensions.kt",
+                    "org/gradle/kotlin/dsl/GeneratedGradleApiExtensionsMainKotlinDslApiExtensionsKt.class")
+                    .map { it.replace('/', File.separatorChar) }.toTypedArray()))
+        }
+    }
+}
+
+
+// See https://github.com/gradle/gradle/blob/master/buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/public-api.kt
+private
+val gradlePublicApiObjectDeclaration = """
```

Comment:
It would probably be good to have a reference back to this file in that one just incase the gradle/gradle one changes.